runtime/io.c

```diff
@@ -401,7 +420,12 @@ void caml_finalize_channel(value vchan)
 {
   struct channel * chan = Channel(vchan);
   if ((chan->flags & CHANNEL_FLAG_MANAGED_BY_GC) == 0) return;
-  if (--chan->refcount > 0) return;
+  unlink_channel(chan);
+  if (atomic_fetch_add (&chan->refcount, -1) > 1) {
+    /* [caml_ml_out_channels_list] may have a reference to this channel. */
+    link_channel (chan);
+    return;
+  }
   caml_plat_mutex_free(&chan->mutex);
   /* TODO KC: See commented out section */
```

Comment:
May or may not be for this PR, but do we want to clean out that commented out section while the code is 'in cache'?